File: market/common-test/src/main/java/ru/yandex/market/common/test/spring/FunctionalTestHelper.java
package ru.yandex.market.common.test.spring;

import org.springframework.http.HttpEntity;
import org.springframework.http.HttpHeaders;
import org.springframework.http.HttpMethod;
import org.springframework.http.MediaType;
import org.springframework.http.ResponseEntity;
import org.springframework.web.client.RequestCallback;
import org.springframework.web.client.ResponseExtractor;
import org.springframework.web.client.RestTemplate;

import javax.annotation.Nullable;
import java.util.Collections;
import java.util.List;
import java.util.Map;

/**
 * Утильный для класс для осуществления запросов к HTTP серверу.
 *
 * @author fbokovikov
 */
public class FunctionalTestHelper {

    private static final RestTemplate REST_TEMPLATE = RestTemplateFactory.createRestTemplate();

    private FunctionalTestHelper() {
        throw new UnsupportedOperationException("Can not instantiate util class");
    }

    public static <T> ResponseEntity<T> get(String url, Class<T> tClass, Object... uriVariables) {
        HttpEntity<String> request = new HttpEntity<>(null, headers());
        return REST_TEMPLATE.exchange(url, HttpMethod.GET, request, tClass, uriVariables);
    }

    public static ResponseEntity<String> get(String url, Object... uriVariables) {
        return exchange(url, null, HttpMethod.GET, uriVariables);
    }

    public static ResponseEntity<String> post(String url) {
        return post(url, (Object) null);
    }

    public static ResponseEntity<String> post(String url, Object body, Object... uriVariables) {
        return exchange(url, body, HttpMethod.POST, uriVariables);
    }

    public static <T> ResponseEntity<String> post(String url, HttpEntity<T> httpEntity) {
        return REST_TEMPLATE.postForEntity(url, httpEntity, String.class);
    }

    public static String postForXml(String url, String request) {
        return REST_TEMPLATE.postForObject(url, new HttpEntity<>(request, xmlHeaders()), String.class);
    }

    public static String postForJson(String url, String body) {
        return REST_TEMPLATE.postForObject(url, new HttpEntity<>(body, jsonHeaders()), String.class);
    }

    public static ResponseEntity<String> put(String url, Object body, Object... uriVariables) {
        return exchange(url, body, HttpMethod.PUT, uriVariables);
    }

    public static <T> void put(String url, HttpEntity<T> httpEntity) {
        REST_TEMPLATE.put(url, httpEntity, String.class);
    }

    public static void putForXml(String url, String request) {
        REST_TEMPLATE.put(url, new HttpEntity<>(request, xmlHeaders()));
    }

    public static void putForJson(String url, String body) {
        REST_TEMPLATE.put(url, new HttpEntity<>(body, jsonHeaders()));
    }

    public static void delete(String url, Object... uriVariables) {
        REST_TEMPLATE.delete(url);
    }

    public static void delete(String url, Map<String, ?> uriVariables) {
        REST_TEMPLATE.delete(url, uriVariables);
    }
    public static ResponseEntity<String> exchange(String url, Object body, HttpMethod method, Object... uriVariables) {
        HttpEntity<Object> request = new HttpEntity<>(body, headers());
        return REST_TEMPLATE.exchange(url, method, request, String.class, uriVariables);
    }

    @Nullable
    public static <T> T execute(String url,
                                HttpMethod method,
                                @Nullable RequestCallback requestCallback,
                                @Nullable ResponseExtractor<T> responseExtractor,
                                Object... uriVariables) {
        return REST_TEMPLATE.execute(url, method, requestCallback, responseExtractor, uriVariables);
    }

    private static HttpHeaders headers() {
        HttpHeaders headers = new HttpHeaders();
        headers.setAccept(Collections.singletonList(MediaType.ALL));
        return headers;
    }

    private static HttpHeaders xmlHeaders() {
        HttpHeaders httpHeaders = new HttpHeaders();
        httpHeaders.setContentType(MediaType.APPLICATION_XML);
        return httpHeaders;
    }

    private static HttpHeaders jsonHeaders() {
        HttpHeaders httpHeaders = new HttpHeaders();
        httpHeaders.setContentType(MediaType.APPLICATION_JSON);
        httpHeaders.setAccept(Collections.singletonList(MediaType.APPLICATION_JSON));
        return httpHeaders;
    }
}

File: market/test/kotlin/ru/yandex/market/adv/incut/utils/FunctionalTestHelper.kt
package ru.yandex.market.adv.incut.utils

import org.springframework.http.HttpEntity
import org.springframework.http.HttpHeaders
import org.springframework.http.HttpMethod
import org.springframework.http.MediaType
import org.springframework.http.ResponseEntity
import org.springframework.web.client.RestTemplate

class FunctionalTestHelper {

    companion object {

        private val REST_TEMPLATE: RestTemplate = RestTemplate()

        fun <T> get(url: String, tClass: Class<T>, vararg uriVariables: Any?): T {
            return getAsEntity(url, tClass, *uriVariables).body
        }

        fun get(url: String): String {
            return get(url, String::class.java)
        }

        fun post(url: String, body: Any? = null, vararg uriVariables: Any?): String {
            return postForEntity(url, body, *uriVariables).body!!
        }

        fun postForEntity(url: String, body: Any? = null, vararg uriVariables: Any?): ResponseEntity<String> {
            val entity = HttpEntity(body, headers())
            return REST_TEMPLATE.exchange(url, HttpMethod.POST, entity, String::class.java, *uriVariables)
        }

        fun put(url: String, body: Any? = null, vararg uriVariables: Any?): String? {
            return putForEntity(url, body, *uriVariables).body
        }

        fun delete(url: String, body: Any? = null, vararg uriVariables: Any?): String? {
            return deleteForEntity(url, body, *uriVariables).body
        }

        fun deleteForEntity(url: String, body: Any? = null, vararg uriVariables: Any?): ResponseEntity<String?> {
            val entity = HttpEntity(body, headers())
            return REST_TEMPLATE.exchange(url, HttpMethod.DELETE, entity, String::class.java, *uriVariables)
        }

        fun putForEntity(url: String, body: Any? = null, vararg uriVariables: Any?): ResponseEntity<String> {
            val entity = HttpEntity(body, headers())
            return REST_TEMPLATE.exchange(url, HttpMethod.PUT, entity, String::class.java, *uriVariables)
        }

        fun <T> getAsEntity(url: String, tClass: Class<T>?, vararg uriVariables: Any?): ResponseEntity<T> {
            val request = HttpEntity<String?>(null, headers())
            return REST_TEMPLATE.exchange(url, HttpMethod.GET, request, tClass, *uriVariables)
        }

        private fun headers(): HttpHeaders {
            val headers = HttpHeaders()
            headers.contentType = MediaType.APPLICATION_JSON
            return headers
        }
    }
}
